src/mono/msvc/shimglobalization.targets

```diff
@@ -0,0 +1,51 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project DefaultTargets="Build" ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <ItemGroup Label="shim_libraries">
+    <ClCompile Include="$(SHIM_GLOBALIZATION)\pal_calendarData.c">
+      <CompileAs>CompileAsCpp</CompileAs>
```

Comment:
Should this really be compiled as C++? By default all our msvc project compiles as C and then there is a global switch to change all to compile using C++ compiler. Only exception to this is LLVM sources that are C++. If this is not the intent, then you can collapse all the ClCompile into one line <ClCompile Include="$(SHIM_GLOBALIZATION)\pal_calendarData.c" />

Reply:
Same for all below.

Reply:
They need to be compiled as c++ otherwise I get a lot of errors like this:
C:\cygwin64\home\thays\runtime\src\libraries\Native\Unix\System.Globalization.Native\pal_icushim_internal.h(165,1): error C2143: syntax error: missing '{' before '*' [C:\cygwin64\home\thays\runtime\src\mono\msvc\libmono-dynamic.vcxproj]

Reply:
does that mean we build these sources with C++ compiler on other platforms as well or that it succeeds compiling using C compiler? If the code claims to be C but fails to compile under C compiler then maybe we should that issue?

Reply:
@safern , this should compile using C compiler?
On linux and mac we are using the default compiler, the same one that is used to compile our sources.

Reply:
I think that the problem is `__typeof` that is used here https://github.com/dotnet/runtime/blob/05954773b6ac44d171dded50e79686619fa71522/src/libraries/Native/Unix/System.Globalization.Native/pal_icushim_internal.h#L164

Reply:
that won't work under msvc when compiling as C (since it's not available). This probably work under gcc since it implements and extension to handle that under C.

Reply:
So it's okay to use `<CompileAs>CompileAsCpp</CompileAs>` or should we try something different?

Reply:
I guess solution would be to explicitly add type as a parameter into PER_FUNCTION_BLOCK in that header to eliminate the dependency on __typeof.

Reply:
@safern what do you think about it?

Reply:
In CoreCLR we build these sources as C++ as well. Do we really want to make this more complicated that what already it is to avoid building as C++? 

We use `PER_FUNCTION_BLOCK` in other places as well:
https://github.com/dotnet/runtime/blob/1db45ed953745c0e270d09df99828cb58772f4ad/src/libraries/Native/Unix/System.Globalization.Native/pal_icushim.c#L22

https://github.com/dotnet/runtime/blob/1db45ed953745c0e270d09df99828cb58772f4ad/src/libraries/Native/Unix/System.Globalization.Native/pal_icushim.c#L322-L324

https://github.com/dotnet/runtime/blob/1db45ed953745c0e270d09df99828cb58772f4ad/src/libraries/Native/Unix/System.Globalization.Native/pal_icushim.c#L334-L336

and then we reference the FOR_ALL_ICU_FUNCTIONS macro in multiple places as well, so we would need to change the way we use `PER_FUNCTION_BLOCK` and the way we define it in all those places to take the type.

Reply:
Main difference is that Mono is a C codebase and CoreCLR a C++ codebase. Complete runtime can be build using C compiler without getting into any risk of taking dependency on C++ runtime, adding size of final binary. I would have switch to setup the function signatures as typedefs and then you can use that in all the calls to PER_FUNCTION_BLOCK instead of the function and rely on __typeof that is not portable. Having that said, as long as we don't introduce any additional C++ runtime dependencies, and that we still can compile as just C on other platforms I believe we could live with compiling this using C++ compiler on Windows. If that turns out to be a problem at a later stage I believe we should refactor and get rid of dependency on non portable feature in ICU shim C source files.

Reply:
If you think that would be a problem at some point I don’t mind changing it to use the type as a macro parameter.